### ISSUE TITLE
GHS compatibility on Windows

### ISSUE THREAD

**Issue by Shionigami** [ccache/ccache#460]
### How to reproduce ###

1. Run ccache on a project using GHS to build (ex: cxrh850.exe)

### Actual behavior ###

ccache is trying to copy some files, defined with the GHS option "-MF", but the file does not exists while ccache is copying it. 

Here is the log which is generated. I also added some debug messages locally, to find where the issue is coming from (I removed most of the messages from the below log). But I was not able to identify the issue.

This leads to thousands of internal errors, and very low performance for ccache.

```
[2019-09-02T14:42:01.416151 5508 ] Command line: c:\Tools\ccache\3.7.3\ccache.exe C:\Tools\greenhills_rel\comp_201815_3fp_x64\cxrh850.exe -MD -MF=D:/GIT/project/rtaos/generatedFiles/dep/uart_workaround.o.d 
[2019-09-02T14:42:39.321534 30044] Working directory: D:\GIT\generatedFiles
[2019-09-02T14:42:39.322530 30044] Source file: D:\GIT\project\tic\src\tic.cpp
[2019-09-02T14:42:39.322530 30044] Dependency file: D:/GIT/generatedFiles/dep/tic.o.d
[2019-09-02T14:42:39.322530 30044] Object file: CMakeFiles\tic.dir\D_\GIT\project\tic\src\tic.o
[2019-09-02T14:42:39.324525 30044] Hashing CWD D:\GIT\generatedFiles
[2019-09-02T14:42:39.325523 30044] Trying direct lookup
[2019-09-02T14:42:39.326549 30044] Looking for object file hash in C:/Users/%USERNAME%/ccache/a60728a/8/b/c3bc7e2dedef0e5c56493b8ab2ee3b-6420.manifest
[2019-09-02T14:42:39.326549 30044] No such manifest file
[2019-09-02T14:42:39.327551 30044] Did not find object file hash in manifest
[2019-09-02T14:42:39.332531 30044] Running preprocessor
[2019-09-02T14:42:40.788250 30044] Got object file hash from preprocessor
[2019-09-02T14:42:40.788250 30044] Preprocessor created D:/GIT/generatedFiles/dep/tic.o.d
[2019-09-02T14:42:40.789248 30044] Object file C:/Users/%USERNAME%/ccache/a60728a/e/e/1c3dba3289960d286b3c352a998fbb-1221786.o not in cache
[2019-09-02T14:42:40.789248 30044] Running real compiler
[2019-09-02T14:42:55.829728 30044] Unlink C:/Users/%USERNAME%/ccache/a60728a/e/e/1c3dba3289960d286b3c352a998fbb-1221786.o.tmp.stdout.LR-Z6379.30044.fhtaht
[2019-09-02T14:42:55.829879 30044] Base dir not set, skip using relative paths
[2019-09-02T14:42:55.829879 30044] Unlink C:/Users/%USERNAME%/ccache/a60728a/e/e/1c3dba3289960d286b3c352a998fbb-1221786.o.tmp.stderr.LR-Z6379.30044.lr7vq8
[2019-09-02T14:42:55.831155 30044] Copying CMakeFiles\tic.dir\D_\GIT\project\tic\src\tic.o to C:/Users/%USERNAME%/ccache/a60728a/e/e/1c3dba3289960d286b3c352a998fbb-1221786.o via C:/Users/%USERNAME%/ccache/a60728a/e/e/1c3dba3289960d286b3c352a998fbb-1221786.o.LR-Z6379.30044.wyhzir (compressed)
[2019-09-02T15:01:16.754474 5508 ] Stored in cache: CMakeFiles\os.dir\D_\GIT\project\os\src\uart_workaround.o -> C:/Users/%USERNAME%/ccache/a60728a/e/4/7600fe9749be948fb6148cc984698c-28467.o (copied)
[2019-09-02T15:01:16.755472 5508 ] Copying D:/GIT/generatedFiles/dep/uart_workaround.o.d to C:/Users/%USERNAME%/ccache/a60728a/e/4/7600fe9749be948fb6148cc984698c-28467.d via C:/Users/%USERNAME%/ccache/a60728a/e/4/7600fe9749be948fb6148cc984698c-28467.d.LR-Z6379.5508.d6zfdd (compressed)
[2019-09-02T15:17:56.812401 5508 ] open error: No such file or directory
[2019-09-02T15:17:56.816373 5508 ] Unlink C:/Users/%USERNAME%/ccache/a60728a/e/4/7600fe9749be948fb6148cc984698c-28467.d.LR-Z6379.5508.d6zfdd
[2019-09-02T15:17:56.816373 5508 ] Failed to copy D:/GIT/generatedFiles/dep/uart_workaround.o.d to C:/Users/%USERNAME%/ccache/a60728a/e/4/7600fe9749be948fb6148cc984698c-28467.d: No such file or directory
[2019-09-02T15:17:56.816373 5508 ] Failed; falling back to running the real compiler
[2019-09-02T15:17:56.816373 5508 ] Executing C:\Tools\greenhills_rel\comp_201815_3fp_x64\cxrh850.exe -MD -MF=D:/GIT/generatedFiles/dep/uart_workaround.o.d
```

### Expected behavior ###
No failure expected. I am not sure what should be the normal behavior for these files.

### Environment ###

ccache 3.7.3
GreenHills compiler 201815
Windows 10

**Comment by peterbud** [comment]
@Shionigami : which version of the ccache did you use:
* self compiled
* MSYS2 version of ccache
* MINGW64 version of ccache

Which shell did you use:
* Windows command prompt
* MSYS2 shell
* MINGW shell

**Comment by Shionigami** [comment]
I am building my own version of ccache, based on 3.7.3, with MinGW64, and running ccache with Windows command prompt.

**Comment by peterbud** [comment]
Currently only the MSYS version of ccache is tested and this only works starting from an MSYS2 shell. The native MINGW64 version is known as not working for a number of test cases.

**Comment by jrosdahl** [comment]
Thanks @peterbud.

Also, I would like to mention that ccache only works with GCC and compilers that behave similar enough, which Green Hills compilers apparently don't. I relabeled this as feature request since I don't consider it to be a bug.

That said, I don't get any real hints from the log file you sent. It's process 5508 that failed, but a) the log also contains irrelevant logs from process 30044 and b) only parts of 5508's logs are present.

Another thing: It's 19 minutes between

```
[2019-09-02T14:42:01.416151 5508 ] Command line: c:\Tools\ccache\3.7.3\ccache.exe C:\Tools\greenhills_rel\comp_201815_3fp_x64\cxrh850.exe -MD -MF=D:/GIT/project/rtaos/generatedFiles/dep/uart_workaround.o.d 
```

and

```
[2019-09-02T15:01:16.754474 5508 ] Stored in cache: CMakeFiles\os.dir\D_\GIT\project\os\src\uart_workaround.o -> C:/Users/%USERNAME%/ccache/a60728a/e/4/7600fe9749be948fb6148cc984698c-28467.o (copied)
[2019-09-02T15:01:16.755472 5508 ] Copying D:/GIT/generatedFiles/dep/uart_workaround.o.d to C:/Users/%USERNAME%/ccache/a60728a/e/4/7600fe9749be948fb6148cc984698c-28467.d via C:/Users/%USERNAME%/ccache/a60728a/e/4/7600fe9749be948fb6148cc984698c-28467.d.LR-Z6379.5508.d6zfdd (compressed)
```

and then it takes another ≈17 minutes for

```
[2019-09-02T15:17:56.812401 5508 ] open error: No such file or directory
[2019-09-02T15:17:56.816373 5508 ] Unlink C:/Users/%USERNAME%/ccache/a60728a/e/4/7600fe9749be948fb6148cc984698c-28467.d.LR-Z6379.5508.d6zfdd
[2019-09-02T15:17:56.816373 5508 ] Failed to copy D:/GIT/generatedFiles/dep/uart_workaround.o.d to C:/Users/%USERNAME%/ccache/a60728a/e/4/7600fe9749be948fb6148cc984698c-28467.d: No such file or directory
[2019-09-02T15:17:56.816373 5508 ] Failed; falling back to running the real compiler
[2019-09-02T15:17:56.816373 5508 ] Executing C:\Tools\greenhills_rel\comp_201815_3fp_x64\cxrh850.exe -MD -MF=D:/GIT/generatedFiles/dep/uart_workaround.o.d
```

to appear. Any ideas why?

**Comment by Shionigami** [comment]
That was because the license server was overloaded, then it took very long to proceed.
Now it takes only a few seconds for each steps.

**Comment by Shionigami** [comment]
I did some investigation about the issue and I think I found the reason why I got problems. 
The only thing is that I am not very familiar with compiler options, then I am not pretty sure about the analysis.

As GHS build is using -MD option, then the boolean "generating_dependencies" is true. 
Same for dependency_filename_specified, as -MF option is used. 

And here is the problem. In my commands, as -MF is used, it shall generate the file registered in the variable "output_dep". 

But, during preprocessor step, the option "-c" and "-o" are removed and replaced:
 "-MQ": does not seems to exist for ARM and GHS compilers. I checked many documentations, and i did not found any info about that. And cxrh850.exe (for GHS) returns a warning, telling that this argument is unknown.
 "-E": Prints the data which shall be written to the file specified with the -MF option are written to the standard output from Windows. Then the file is not generated.

The next step is to copy this file (specified with -MF) to the cache, but, as it is not generated, ccache fails.

Then, what could be the fix? 
I just did test and am still using "-c" and "-o" options for preprocessor steps, it is working well, but I am pretty sure that this is not the way it should be used.

**Comment by Shionigami** [comment]
I created a PR here with the changes:
https://github.com/ccache/ccache/pull/471

For comparison, here are the results with and without my changes:
3.7.3 without changes: 
`cache directory                     C:/Users/%USERNAME%/ccache/5d6f2cb
primary config                      C:/Users/%USERNAME%/ccache/5d6f2cb/ccache.conf
secondary config      (readonly)    /usr/local/etc/ccache.conf
stats updated                       09/12/19 16:00:40
stats zeroed                        09/12/19 14:30:28
cache hit (direct)                     0
cache hit (preprocessed)               0
cache miss                             4
cache hit rate                      0.00 %
ccache internal error               1737
cleanups performed                     0
files in cache                      2422
cache size                          71.0 MB
max cache size                       2.0 GB`

3.7.3 with updates (First run)
`cache directory                     c:\users\%USERNAME%\.ccache
primary config                      c:\users\%USERNAME%\.ccache/ccache.conf
secondary config      (readonly)    /usr/local/etc/ccache.conf
cache hit (direct)                     0
cache hit (preprocessed)               0
cache miss                             0
cache hit rate                      0.00 %
cleanups performed                     0
files in cache                         0
cache size                           0.0 kB
max cache size                       5.0 GB
With ccache --- 3676.11999989 seconds ---`

3.7.3 with updates (second run)
`cache directory                     C:/Users/%USERNAME%/ccache/a60728a
primary config                      C:/Users/%USERNAME%/ccache/a60728a/ccache.conf
secondary config      (readonly)    /usr/local/etc/ccache.conf
stats updated                       09/12/19 14:35:06
stats zeroed                        09/12/19 14:31:43
cache hit (direct)                  1739
cache hit (preprocessed)               1
cache miss                             0
cache hit rate                    100.00 %
cleanups performed                     0
files in cache                      5900
cache size                          76.2 MB
max cache size                       5.0 GB
With ccache --- 399.994999886 seconds ---`

**Comment by jrosdahl** [comment]
Note: As of ccache version 3.7.7 the `-MQ` option is now again used. See 6d4537695f8753556362c6c5521360ca9b596ab2 for why.

**Comment by jrosdahl** [comment]
> Note: As of ccache version 3.7.7 the -MQ option is now again used. See 6d45376 for why.

Another solution for not using -MQ has now been implemented: 07b55f13e119a2a35d166a73929643911ea69d52 (bound for ccache 4.0).
